sound/soc/sof/intel/hda-dai.c

```diff
@@ -475,17 +480,78 @@ static int ssp_dai_hw_params(struct snd_pcm_substream *substream,
 			     struct snd_pcm_hw_params *params,
 			     struct snd_soc_dai *dai)
 {
+	struct ssp_dai_dma_data *dma_data;
+
+	dma_data = kzalloc(sizeof(*dma_data), GFP_KERNEL);
+	if (!dma_data)
+		return -ENOMEM;
+
+	snd_soc_dai_set_dma_data(dai, substream, dma_data);
+
 	return ssp_dai_setup_or_free(substream, dai, true);
```

Comment:
Should we set the dma_data only if ssp_dai_setup_or_free() returned without error?

Reply:
dunno. If hw_params fails, is hw_free called?

Reply:
move to startup